cuML/src/common/device_buffer.hpp

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "../../../ml-prims/src/utils.h"
+
+#include "../cuML.hpp"
+
+namespace ML {
+
+/**
+ * RAII object owning a contigous typed device buffer. The passed in allocator supports asynchronus allocation and
+ * deallocation so this can be used for temporary memory 
+ * @code{.cpp}
+ * void foo( cumlHandle* handle, .., cudaStream_t stream )
+ * {
+ *     ...
+ *     device_buffer<T> temp( handle->getHostAllocator(), 0 )
+ *     
+ *     temp.resize(n, stream);
+ *     kernelA<<<grid,block,0,stream>>>(...,temp.data(),...);
+ *     kernelB<<<grid,block,0,stream>>>(...,temp.data(),...);
+ *     temp.release(stream);
+ * }
+ * @endcode
+ */
+template<typename T>
+class device_buffer
+{
+public:
+    typedef std::size_t         size_type;
+    typedef T                   value_type;
+    typedef value_type*         iterator;
+    typedef const value_type*   const_iterator;
+    typedef T&                  reference;
+    typedef const T&            const_reference;
+
+    device_buffer() = delete;
+
+    device_buffer(const device_buffer& other) = delete;
+
+    device_buffer& operator=(const device_buffer& other) = delete;
+
+    device_buffer(std::shared_ptr<deviceAllocator> allocator, size_type n = 0)
+        : _allocator(allocator), _size(n), _capacity(n), _data(0)
+    {
+        if ( n > 0 )
+        {
+            _data = static_cast<value_type*>(_allocator->allocate( _capacity*sizeof(value_type), 0 ));
+            CUDA_CHECK( cudaStreamSynchronize( 0 ) );
+        }
+    }
+
+    ~device_buffer()
+    {
+        if ( 0 != _data ) 
+        {
+            _allocator->deallocate( _data, _capacity*sizeof(value_type), 0 );
```

Comment:
Same as above comment for the ctor method. default stream sync is being imposed?

Reply:
See my reply there. I have choose this as the alternative with the least surprises for the user. If asynchronous free is required `release` that way it is transparent to the user of device_buffer that the buffer is freed asynchronously.